types/decimal.go

```diff
@@ -0,0 +1,422 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const (
+	Precision = 10
+
+	// bytes required to represent the above precision
+	// ceil(log2(9999999999))
+	DecimalPrecisionBytes = 34
+)
+
+var (
+	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	fivePrecision        = new(big.Int).Quo(precisionReuse, big.NewInt(2))
+	precisionMultipliers []*big.Int
+	zeroInt              = big.NewInt(0)
+	oneInt               = big.NewInt(1)
+	tenInt               = big.NewInt(10)
+)
+
+// Set precision multipliers
+func init() {
+	precisionMultipliers = make([]*big.Int, Precision+1)
+	for i := 0; i <= Precision; i++ {
+		precisionMultipliers[i] = calcPrecisionMultiplier(int64(i))
+	}
+}
+
+func precisionInt() *big.Int {
+	return new(big.Int).Set(precisionReuse)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{new(big.Int).Set(zeroInt)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// calculate the precision multiplier
+func calcPrecisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic(fmt.Sprintf("too much precision, maximum %v, provided %v", Precision, prec))
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(tenInt, big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// get the precision multiplier. Do not mutate result.
```

Comment:
Maybe a way to make Dec immutable so we can enforce it?

Reply:
These are only used within this file, so we don't need to add a separate method to enforce all decimals are immutable.

Reply:
leaving for now